File: agent/tools/tool_search/tool_search_tool.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

from llm.tool_spec import ToolSpec
from tools.base_tool import BaseTool
from tools.tool_permission import ToolPermission
from tools.tool_result import ToolResult


class ToolSearchInput(BaseModel):
    keywords: str = Field(..., min_length=1, description="空格分隔的搜索关键词")
    max_results: int = Field(default=3, ge=1, le=10)


class ToolSearchTool(BaseTool[ToolSearchInput, BaseModel]):
    """让模型按关键词搜索并发现延迟加载的工具。"""

    def __init__(self, specs_provider: Any) -> None:
        super().__init__(
            name="tool_search",
            description="按关键词搜索可用的延迟加载工具，获取其完整输入参数 schema。",
            input_model=ToolSearchInput,
            required_permissions={ToolPermission.LLM},
            category="meta",
        )
        self._specs_provider = specs_provider
        self._is_read_only = True
        self._is_concurrency_safe = True
        self._should_defer = False
        self._always_load = True
        self._search_hint = "工具,搜索,发现,加载"

    async def execute(self, tool_input: ToolSearchInput, context: dict[str, Any]) -> ToolResult:
        _ = context
        keywords = [kw.strip().lower() for kw in tool_input.keywords.split() if kw.strip()]
        if not keywords:
            return ToolResult.error("tool_search: empty keywords")

        candidates: list[ToolSpec] = self._specs_provider() if callable(self._specs_provider) else []
        scored: list[tuple[int, ToolSpec]] = []
        for spec in candidates:
            if spec.name == "tool_search":
                continue
            text = f"{spec.name} {spec.description} {spec.search_hint}".lower()
            score = sum(1 for kw in keywords if kw in text)
            if score > 0:
                scored.append((score, spec))

        scored.sort(key=lambda item: item[0], reverse=True)
        top = scored[: tool_input.max_results]

        if not top:
            return ToolResult(
                ok=True,
                status="miss",
                message="未找到匹配的工具，尝试更换关键词",
                items=[],
            )

        items = []
        for _, spec in top:
            schema_text = (
                f"工具名称: {spec.name}\n"
                f"描述: {spec.description}\n"
                f"输入参数 (JSON Schema):\n{spec.parameters}"
            )
            items.append({
                "tool_name": spec.name,
                "description": spec.description,
                "parameters": spec.parameters,
                "schema_text": schema_text,
            })

        return ToolResult(
            ok=True,
            status="hit",
            message=f"找到 {len(items)} 个匹配工具",
            items=items,
        )

File: agent-ts/src/tools/search/ToolSearchTool.ts
import type { JsonObject } from "../../common/json/types/JsonTypes.js";
import type { OpenAIChatTool } from "../../openai/chat/model/tool/OpenAIChatTool.js";

export interface ToolSearchSpec {
  readonly name: string;
  readonly description: string;
  readonly parameters: JsonObject;
  readonly searchHint?: string;
}

export class ToolSearchTool {
  constructor(private readonly specsProvider: () => Promise<readonly ToolSearchSpec[]> | readonly ToolSearchSpec[]) {}

  create(): OpenAIChatTool {
    return {
      type: "function",
      function: {
        name: "tool_search",
        description: "按关键词搜索可用的延迟加载工具，获取其完整输入参数 schema。",
        parameters: {
          type: "object",
          properties: {
            keywords: { type: "string" },
            max_results: { type: "number" }
          },
          required: ["keywords"]
        }
      }
    };
  }

  async execute(keywords: string, maxResults = 3): Promise<JsonObject> {
    const tokens = keywords
      .split(/\s+/u)
      .map((item) => item.trim().toLowerCase())
      .filter(Boolean);
    if (tokens.length === 0) {
      return { ok: false, status: "error", message: "tool_search: empty keywords", items: [] };
    }

    const candidates = [...(await this.specsProvider())];
    const scored: Array<[number, ToolSearchSpec]> = [];
    for (const spec of candidates) {
      if (spec.name === "tool_search") continue;
      const text = `${spec.name} ${spec.description} ${spec.searchHint ?? ""}`.toLowerCase();
      const score = tokens.reduce((total, token) => total + (text.includes(token) ? 1 : 0), 0);
      if (score > 0) scored.push([score, spec]);
    }

    scored.sort((left, right) => right[0] - left[0]);
    const top = scored.slice(0, Math.max(1, Math.min(10, maxResults)));
    if (top.length === 0) {
      return {
        ok: true,
        status: "miss",
        message: "未找到匹配的工具，尝试更换关键词",
        items: []
      };
    }

    return {
      ok: true,
      status: "hit",
      message: `找到 ${top.length} 个匹配工具`,
      items: top.map(([, spec]) => ({
        tool_name: spec.name,
        description: spec.description,
        parameters: spec.parameters,
        schema_text: `工具名称: ${spec.name}\n描述: ${spec.description}\n输入参数 (JSON Schema):\n${JSON.stringify(spec.parameters, null, 2)}`
      }))
    };
  }
}
